lib/fluri.dart

```diff
@@ -199,16 +199,71 @@ class FluriMixin {
     _uri = _uri.replace(queryParameters: queryParameters);
   }
 
+  /// The URI query parameters with support for multi-value params.
+  Map<String, List<String>> get queryParametersAll => _uri.queryParametersAll;
+  set queryParametersAll(Map<String, Iterable<String>> queryParameters) {
+    _uri = _uri.replace(queryParameters: queryParameters);
+  }
+
   /// Set a single query parameter.
-  void setQueryParam(String param, String value) {
+  ///
+  /// If the given query parameter name is already set, it will be completely
+  /// replaced with the given [value].
+  ///
+  /// Throws an [ArgumentError] if [value] is not a `String` or an
+  /// `Iterable<String>`.
+  void setQueryParam(String param, dynamic /*String|Iterable<String>*/ value) {
+    if (value is! String && value is! Iterable<String>) {
+      throw new ArgumentError.value(
+          value,
+          'value',
+          'Must be a String or '
+          'Iterable<String>');
+    }
+
     updateQuery({param: value});
   }
 
   /// Update the URI query parameters, merging the given map with the
   /// current query parameters map instead of overwriting it.
-  void updateQuery(Map<String, String> queryParameters) {
-    var newQueryParameters = new Map<String, String>.from(this.queryParameters);
-    newQueryParameters.addAll(queryParameters);
+  ///
+  /// Throws an [ArgumentError] if any value in the given
+  /// [queryParametersToUpdate] map is not a `String` or an `Iterable<String>`.
+  ///
+  /// If [mergeValues] is `false`, each parameter in [queryParametersToUpdate]
+  /// will be completely replaced with the new value.
+  ///
+  /// If [mergeValues] is `true`, the values for each parameter in
+  /// [queryParametersToUpdate] will be merged into the existing list of values
+  /// for that parameter. Duplicate values will be discarded.
+  void updateQuery(
+      Map<String, dynamic /*String|Iterable<String>*/ > queryParametersToUpdate,
+      {bool mergeValues: false}) {
+    var newQueryParameters = <String, List<String>>{};
+
+    void _updateQueryParams(
+        Map<String, dynamic /*String|Iterable<String>*/ > qp) {
+      qp.forEach((key, value) {
+        if (!mergeValues || !newQueryParameters.containsKey(key)) {
+          newQueryParameters[key] = [];
+        }
+        if (value is String && !newQueryParameters.containsValue(value)) {
+          newQueryParameters[key].add(value);
+        } else if (value is Iterable<String>) {
+          for (var v in value) {
+            if (!newQueryParameters[key].contains(v)) {
+              newQueryParameters[key].add(v);
+            }
+          }
+        } else {
+          throw new ArgumentError('Query parameter "$key" has value "$value" '
+              'which is not a String or an Iterable<String>.');
+        }
+      });
+    }
+
+    _updateQueryParams(this.queryParametersAll);
```

Comment:
this seems unnecessary - `queryParametersAll` contains your starting point, so the `_updateQueryParams` logic is mostly unnecessary right?  could just copy it into `newQueryParameters`?